### DISCUSSION TITLE
Is there a way to output multiple items in a single message?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [jmcdo29/ogma#1381] [Q&A]
When we run `console.log('some data:', object)` it outputs `some data:`, a space and the `object` in ‘JavaScript format’ (i.e. not in valid JSON format, but similar to it).

Now I’d like to use Ogma similarly, i.e. to output two or more items in one call. I understand that we cannot do it same way as we do it with `console.log()`, however, could we use, e.g., an array as the first parameter of Ogma logger methods, potentially with a flag/property in the Ogma options object (you call it `meta`) to make Ogma output the array items individually, rather than as an array?

I would be okay if any object/array would be output as a valid JSON.

**Comment by jmcdo29** [comment]
Currently, this functionality doesn't exist. However, I think we could take the `meta` object and add an `each` property to it, and if `true`, then iterate over the array as you mentioned. I'll try to get this whipped up this week

**Reply by tukusejssirs** [reply]
Thanks! :pray:

**Reply by tukusejssirs** [reply]
I have just had a change to test this out. It works, but not exactly as I would expect.

```ts
// With newline at the end an array item: don't add anything between array items
// Note: `console.log()` actually adds a space between array items nonetheless, however. You might decide to add it (in order to be consistent with `console.log()`), however, I think it is redundant and noisy as the added space is the first character on the line.
logger.error(['some string:\n', {a: 1, b: 2}], {each: true})

// Actual output
[2022-10-05T07:19:06.214Z] [ERROR] [host]  [232613]   some string:

[2022-10-05T07:19:06.215Z] [ERROR] [host]  [232613]
{
  "a": 1,
  "b": 2
}

// Expected output
[2022-10-05T07:19:06.214Z] [ERROR] [host]  [232613]   some string:
{
  "a": 1,
  "b": 2
}

/////////////

// Without at the end an array item: add a single space between array items.
logger.error(['some string:', {a: 1, b: 2}], {each: true})

// Actual output
[2022-10-05T07:19:06.214Z] [ERROR] [host]  [232613]   some string:
[2022-10-05T07:19:06.215Z] [ERROR] [host]  [232613]
{
  "a": 1,
  "b": 2
}

// Expected output
[2022-10-05T07:19:06.214Z] [ERROR] [host]  [232613]   some string: {
  "a": 1,
  "b": 2
}
```

**Reply by tukusejssirs** [reply]
Just a note: the current output does not makes sense, as Ogma users can already do that without `{each: true}`, albeit with multiple commands.

**Reply by jmcdo29** [reply]
Ah, I see. Then I misunderstood the request.

> the current output does not makes sense, as Ogma users can already do that without {each: true}, albeit with multiple commands.

That was actually something I mentioned in the changelog :laughing: 

All right, I'll get back to this and see what can be done. Quick question: how would this effect JSON logging? Would it just be a single message to keep all the metadata consolidated, or would it need to be a full JSON log for each message?

**Reply by tukusejssirs** [reply]
> Ah, I see. Then I misunderstood the request.

I should’ve been more explicit in the first place. However, you could have asked me to test the PR after you created it and before merging it too. :wink:

> Quick question: how would this effect JSON logging? Would it just be a single message to keep all the metadata consolidated, or would it need to be a full JSON log for each message?

Good question. I haven’t thought about that as I haven’t used Ogma JSON logging yet.

As I see it, it is a single log command, I have always considered the output of `console.log(1, 2, 3, ...)` as a single log message, therefore I suggest to treat it as a single JSON log too. If a user wants to create multiple (JSON) logs, they should use multiple commands IMO.

Regarding the JSON log message structure/type, I think it should be always the same as user defines in the first parameter (`message`) of Ogma log command.

**Reply by jmcdo29** [reply]
> However, you could have asked me to test the PR after you created it and before merging it too. wink

I mean, I could've, or I can always release a `patch` to fix the functionality ;)

> As I see it, it is a single log command, I have always considered the output of console.log(1, 2, 3, ...) as a single log message, therefore I suggest to treat it as a single JSON log too

Cool, so this should really only be a "stream" message change then :+1:

> As I see it, it is a single log command, I have always considered the output of console.log(1, 2, 3, ...) as a single log message, therefore I suggest to treat it as a single JSON log too

Yeah, pretty much whatever is passed as the first parameter in a JSON format will be the `message` value

**Reply by jmcdo29** [reply]
@tukusejssirs is this more of what you're looking for?

```ts
const messages = ['hello', 42, { key: 'value' }, true];
ogma.log(messages, { each: true });
```
produces
```
[2022-10-05T21:45:01.372Z] [INFO]  [pop-os]  [3164676]   hello
42
{
  "key": "value"
}
true
```

**Reply by tukusejssirs** [reply]
Yeah. it’s better, but we’re not there yet.

You join the items with a newline, however, `console.log()` joins them with a space.

One thing I don‘t like is addition of a space _after a newline_ (last character of an item in the message), i.e. `console.log('text:\n', {a: 1}) should not add a space between the items.

Could you please check again the examples I wrote in [here](https://github.com/jmcdo29/ogma/discussions/1381#discussioncomment-3802352) please? :pray:

**Reply by jmcdo29** [reply]
Okay, I getcha. You'd be more looking for 
```
[2022-10-05T21:45:01.372Z] [INFO]  [pop-os]  [3164676]   hello 42 { "key": "value" } true
```
My only concern is how are complex JSON objects treated e.g: 
```
console.log("Hello", 'world', { deep: { nest: { key: 'value' }, nest2: {key: 'value' } }, deep2: { nest: { value: 'key'}, nest2: { key: 'value' } } }, true, 42);
```
results in 
```
Hello world {
  deep: { nest: { key: 'value' }, nest2: { key: 'value' } },
  deep2: { nest: { value: 'key' }, nest2: { key: 'value' } }
} true 42
```
wheras
```
 console.log("Hello", 'world', { deep: { nest: { key: 'value' } }, deep2: { nest: { value: 'key'} } }, true, 42);
```
results in
```
Hello world { deep: { nest: { key: 'value' } }, deep2: { nest: { value: 'key' } } } true 42
```

---

I wonder what the criteria for Node are for handling this? Personally, I'd like a blanket rule, my opinion is to just use `JSON.stringify(obj, replacer, 2)` and make the logs come out as
```
[2022-10-05T21:45:01.372Z] [INFO]  [pop-os]  [3164676]   hello 42 {
"key": "value"
} true
```

Which I think would suffice with your expected/actuals above

**Reply by tukusejssirs** [reply]
> Okay, I getcha. You'd be more looking for 
> ```
> [2022-10-05T21:45:01.372Z] [INFO]  [pop-os]  [3164676]   hello 42 { "key": "value" } true
> ```

Not on a single line.

> I wonder what the criteria for Node are for handling this? Personally, I'd like a blanket rule, my opinion is to just use `JSON.stringify(obj, replacer, 2)` and make the logs come out as
> ```
> [2022-10-05T21:45:01.372Z] [INFO]  [pop-os]  [3164676]   hello 42 {
> "key": "value"
> } true
> ```
> 
> Which I think would suffice with your expected/actuals above

By _I would be okay if any object/array would be output as a valid JSON_ in the discussion description I meant exactly `JSON.stringify(o, replacer, 2)`. I should’ve been much clearer then. :wink:

No need to imitate the object ouput of `console.log()` (i.e. without quotes around props or making a decision whether to output the object on a single line). :wink:

**Reply by jmcdo29** [reply]
Quotes around props comes in as part of JSON stringification, and I'm not gonna re-invent the wheel around that, but the rest of this is a simple switch of the `join()` parameter and updating the test
![image](https://user-images.githubusercontent.com/28268680/194191915-c40eabb8-9725-427b-9a61-1b4bd248c634.png)
```
[2022-10-06T01:07:48.661Z] [INFO]  [pop-os]  [3255819]   hello 42 {
  "key": "value"
} true
```

**Reply by tukusejssirs** [reply]
Seems to be perfect. :pray:

Just a question: do you plan removing a space (between array items) when the last character of the previous item is a newline? I understand that it is an extra condition, however, I think the output will look better. :wink:

**Reply by jmcdo29** [reply]
The logic there actually isn't too bad. `message = message.join(' ').replace(/\n\s([\S])/g, '\n$1');` does the trick. Joins all the messages after they're stringified (if they're objects) with a single space and then removes any `\n<space>` so long as it's followed by a non-space character. This makes sure that the stringified objects still keep their formatting. I'm gonna need to add in some tests around that to verify though. I'll probably put this patch up later today.

**Reply by tukusejssirs** [reply]
> `message = message.join(' ').replace(/\n\s([\S])/g, '\n$1');`

What about a situation when the user actually wants to add an explicit whitespace after a `\n`? What if an array item contains something like `asdf\n\tasdf` or `asdf\n{\n  a: 1\n}`?

I think that the replacing should occur _before_ joining the array.

**Reply by jmcdo29** [reply]
> I think that the replacing should occur before joining the array.

Maybe I'm not quite following here. What replacement would happen before the join? There's no extra space at the beginning of the line until the `messages.join(' ')` happens, so I'm not sure what is expected here. 

Do you think you could get me a few test arrays with what you think the expected output would be?

**Reply by tukusejssirs** [reply]
Here you go.

However, I am not sure I we want to treat trailing `\t` in array items similarly as `\n` (i.e. whether we should not add a space after it). I don’t usually use literal tabs in logs, however, it you thing it is a good idea, simply replace:

```
// Find
output[mI] += `${vS}${/[^\n]$/.test(vS) && i + 1 < a.length ? ' ' : ''}`
// Replace
output[mI] += `${vS}${/[^\n\t]$/.test(vS) && i + 1 < a.length ? ' ' : ''}`
```

Note that I simply copied all test arrays from this discussion and used them for the tests below without cleaning them up. You might want to clean them up. :wink:

```ts
const messages = [
  ['asdf\n', 'sdfg', 'werr'],
  ['some string:\n', {a: 1, b: 2}],
  ['some string:', {a: 1, b: 2}],
  [1, 2, 3],
  ['hello', 42, {key: 'value'}, true],
  ["Hello", 'world', {deep: {nest: {key: 'value'}, nest2: {key: 'value'}}, deep2: {nest: {value: 'key'}, nest2: {key: 'value'}}}, true, 42],
  ["Hello", 'world', {deep: {nest: {key: 'value'}}, deep2: {nest: {value: 'key'}}}, true, 42],
  ['asdf\n\tasdf', 'asdf\n{\n a: 1\n}']
]

const expected = [
  'asdf\nsdfg werr',
  'some string:\n{\n  "a": 1,\n  "b": 2\n}',
  'some string: {\n  "a": 1,\n  "b": 2\n}',
  '1 2 3',
  'hello 42 {\n  "key": "value"\n} true',
  'Hello world {\n  "deep": {\n    "nest": {\n      "key": "value"\n    },\n    "nest2": {\n      "key": "value"\n    }\n  },\n  "deep2": {\n    "nest": {\n      "value": "key"\n    },\n    "nest2": {\n      "key": "value"\n    }\n  }\n} true 42',
  'Hello world {\n  "deep": {\n    "nest": {\n      "key": "value"\n    }\n  },\n  "deep2": {\n    "nest": {\n      "value": "key"\n    }\n  }\n} true 42',
  'asdf\n\tasdf asdf\n{\n a: 1\n}'
]

const output = Array(messages.length).fill('')

// Run tests
messages.forEach((m, mI) => {
  m.forEach((v, i, a) => {
    const vS = typeof v === 'object' ? JSON.stringify(v, null, 2).toString() : v.toString()
    output[mI] += `${vS}${/[^\n]$/.test(vS) && i + 1 < a.length ? ' ' : ''}`
  })
})

// Output tests outcome
expected.forEach((x, i) => {
  console.log(`[${i}] did test passed? : ${output[i] === x}`)

  if (output[i] !== x) {
    console.log(`[${i}] : out :`, output[i])
    console.log(`[${i}] : exp :`, x)
  }
})
```

**Reply by jmcdo29** [reply]
Awesome! This expected array helps out a lot. And thanks for adding in a possible solution! One last question, what should happen if something like `['goodbye\n', 'world']` is passed? Should it come out as 
```
<log prefix stuff> goodbye
 world
```
or as 
```
<log prefix stuff> goodbye
world
```
I think the former matches more closely to your request, it's just one of those weird cases, y'know?

**Reply by tukusejssirs** [reply]
At first, I didn’t get why would you consider this a different case from `['asdf\n', 'sdfg']`, however, now I think you treat `world` as a signature (like in a letter or an email).

While it might be different from in signatures, however, who would ever provide signatures in logs? IMHO it won’t happen in real world, thus I presume it is enough to treat it the same way as `['asdf\n', 'sdfg']`.

If I haven’t understood your reason on `['goodbye\n', 'world']` being different from `['asdf\n', 'sdfg']`, explain it to me please. :wink:

**Reply by jmcdo29** [reply]
Somehow I missed seeing the `['asdf\n', 'sdfg']`, I was reading it as a single string in my mind, so the output
```
asdf
sdfg
```
Made sense. (this is the output in the expected array, which I'm actually very happy with).

Right now, the only test that is failing is this one: `['asdf\n\tasdf', 'asdf\n{\n a: 1\n}']` -> `'asdf\n\tasdf asdf\n{\n a: 1\n}'`. This feels like such a contrived edge case that I'm not sure it's ever going to show up. It's hard to say how it should act otherwise though, if I wanted to just keep the `message.join(' ')` as I already have. I could replace spaces with a space character (`\x20`), but that feels like a hack.

**Reply by tukusejssirs** [reply]
It seems to me you don’t like my solution for some reason. May I know why? :wink:

**Reply by jmcdo29** [reply]
Just don't want to add in a for loop. The `join` method is also slightly faster than a for loop is (at least it was in the small benchmark I was doing). I could definitely throw the for loop in there, really just wanted to keep the code looking simple. This is the current implementation
```ts
    if (Array.isArray(message) && each) {
      for (let i = 0; i < message.length; i++) {
        message[i] = this.stringifyObject(message[i], i === 0);
      }
      message = message.join(' ').replace(/\n ([^ ])/g, '\n$1');
    } else {
      message = this.stringifyObject(message);
    }
```
I'd have to fix the `message.join(' ').replace()` call to be something like
```ts
    if (Array.isArray(message) && each) {
      for (let i = 0; i < message.length; i++) {
        message[i] = this.stringifyObject(message[i], i === 0);
      }
      let trueStr = '';
      for (let i = 0; i < message.length; i++) {
        trueStr += `${message[i]}${/[^\n]$/.test(message[i]) && i + 1 < message.length ? ' ' : ''}`
      }
      message = trueStr;
    } else {
      message = this.stringifyObject(message);
    }
```

Actually, thinking about this, I should be able to move the addition of the space up into the `stringifyObject` call, and then just `join('')` as everything should already have the space added if necessary and we just need to take the array to a single sting. That way, we don't have to loop over the array twice (not efficient at all) and we can just pass a condition of if a space should be added based on index to the method, while doing the test in the method itself.

**Reply by jmcdo29** [reply]
Also! Saw that _my_ example was actually wrong. `['goodbye\n' 'world']` was _supposed_ to be `['goodbye\n world']`, single string with a newline **and** a space. If we wanted to keep the formatting of it, with the `join(' ').replace(...)` it's difficult. With the idea you gave me above, it's trivial.

**Reply by tukusejssirs** [reply]
> Actually, thinking about this, I should be able to move the addition of the space up into the `stringifyObject` call, and then just `join('')` as everything should already have the space added if necessary and we just need to take the array to a single sting. That way, we don't have to loop over the array twice (not efficient at all) and we can just pass a condition of if a space should be added based on index to the method, while doing the test in the method itself.

I’d suggest that myself. :wink:

> `['goodbye\n' 'world']` was supposed to be `['goodbye\n world']`

I that case it is similar to my case of `['asdf\n\tasdf', 'asdf\n{\n a: 1\n}']` (the last item).

**Reply by jmcdo29** [reply]
Cool. Updating the test suite to look better, everything is passing though. I'll throw up some sample output soon and let you know when the PR is up so you can take a look yourself. Thanks for that great discussion around this

**Reply by tukusejssirs** [reply]
Thanks, @jmcdo29! :pray:

> Thanks for that great discussion around this

I thank _you_! I feel at least a bit helpful, albeit it is actually you who help me. :wink:

**Comment by tukusejssirs** [comment]
@jmcdo29, I have been using `each` for a quite while and I noticed that I prefer using it by default (i.e. without appending `{each: true}` to each log command).

Would it be possible to define the default value of `each`, e.g., in the Nest module (either in `OgmaModule.forRoot()` or `OgmaModule.forFeature()`) and execute `this.logger.info([1, 2, 3])`?

Thanks! :pray:

**Reply by jmcdo29** [reply]
Looking through the code, yeah I think we can make `each` an option that can be passed to the logger (and `forRoot`) so that you don't have to pass it on each method call. I'll try to implement that this week

**Reply by humblecoder** [reply]
Any update on this? More flexible forms of "structured logging" would be nice.

**Reply by jmcdo29** [reply]
Oh man, sorry, it seems this slipped my mind! Let me make this an issue so I can not lose track of it again

https://github.com/jmcdo29/ogma/issues/1521

**Reply by tukusejssirs** [reply]
@jmcdo29, I didn’t wanted to urge you. :wink:

@humblecoder, it is good to know that I am not the only one to use this feature. :wink: BTW, what do you mean by _more flexible forms of "structured logging"_? This `{each: true}` feature? Or do you have something even even more flexible in mind?

**Reply by humblecoder** [reply]
@tukusejssirs honestly, **ultimately** I'm looking to organize my logged data for visualization of samples/datapoints (think DataDog).  But I want to start with a quality, **speedy**, foundation and "log routing" mechanism.

**Reply by tukusejssirs** [reply]
@humblecoder, isn’t the JSON output format of Ogma logger better for processing by log analyser? I use `{each: true}` to pretty-print all items in an array, similarly (not exactly) to how `console.log(...items)` (e.g. `console.log(1, 2, 3)`) outputs the data, esp objects and array. In JSON output, the log message structure will be same as input, no matter what is the value of `each`.

Maybe I misunderstood you. :thinking:

**Reply by humblecoder** [reply]
It's quite possible that *I* misunderstand.  Yes, I use JSON, but I was looking for the ability to cleanly "multi-output" JSON objects.  Currently they all show up as a kluged string value.  Of course, ultimately that could probably more readily be resolved with my tooling (e.g. SigNoz, Grafana/Loki) than Ogma.

**Reply by tukusejssirs** [reply]
> Yes, I use JSON, but I was looking for the ability to cleanly "multi-output" JSON objects.

If by _cleanly_ you mean _pretty-print_ (like `JSON.stringify(o, null, 2)`), AFAIK that is not an option when using JSON output. JSON output is IMHO intentionally minified. It should be processed by a log analyser/visualiser.

If you want to pretty-print some log messages, you should not use JSON output (`json` set to `false`).

Ultimately, @jmcdo29 has more knowledge if it is possible to achieve what you what.

**Reply by jmcdo29** [reply]
Just wanting to mention that if you want `JSON` logs but still want it to print out in a "clean" manner (kinda like `pretty-print` with pino) you can use the `@ogma/cli` like `nest start --watch | ogma` so that it translates the `JSON` into ogma's string format.

I do think I still have a bug with it and multiple `@nestjs/common` installs due to how `nest-commander` works through, so be warned about that (and let me know if it's still an issue that I need to get back to fixing)

**Reply by tukusejssirs** [reply]
I didn’t know about that, @jmcdo29. Not that I need such feature, but I didn’t read about it anywhere. Is it documented? :thinking:

**Reply by jmcdo29** [reply]
> I didn’t know about that, @jmcdo29. Not that I need such feature, but I didn’t read about it anywhere. Is it documented? thinking

Shoot, looks like I forgot to add that functionality to the docs. Just another thing to update :laughing: There is a mention of it in a [CHANGELOG](https://github.com/jmcdo29/ogma/blob/main/packages/cli/CHANGELOG.md#210) though

**Reply by tukusejssirs** [reply]
I am using Ogma logger since August 2022 (`v2.1.0` was released on [4 Nov 2021](https://github.com/jmcdo29/ogma/releases/tag/%40ogma%2Fcli%402.1.0)), therefore I haven’t read changelogs / release logs before that that. :wink:
